assets/test_chambers/test_chamber_09/test_chamber_09.yaml

```diff
@@ -23,5 +25,20 @@ cutscenes:
     - play_animation pit_1 pit_1_down -1
     - delay 3
     - goto pit_loop
+  CHECK_TRAPPED:
+    - wait_for_signal cube_on_piston 210
+    - wait_for_signal not_escape_hatch
+    - q_sound 08_PART1_TRAPPED_1 CH_GLADOS PORTAL_08_PART1_TRAPPED_1
+    - q_sound 08_PART1_TRAPPED_2 CH_GLADOS PORTAL_08_PART1_TRAPPED_2
+    - wait_for_channel CH_GLADOS
+    - set_signal escape_hatch
+  CHECK_DOOR_EXPLODE:
+    - wait_for_signal open_door_cube
+    - wait_for_signal not_open_door_cube
```

Comment:
In the original game the door explosion happens only when the cube is moved a bit away from the button (there is a trigger around it). That way, the door won't explode if the player is just fumbling to put the cube on the button and removes it temporarily. They have to actually take it away from the button's location.

Can you add this trigger as well and use it instead of `not_open_door_cube`? The YAML in my other comment shows this.

Reply:
While implementing the changes I came across a bug in the button code.

```
if ((other->body->flags & RigidBodyFlagsGrabbable) == RigidBodyFlagsGrabbable && (other->body->sleepFrames <= IDLE_SLEEP_FRAMES - 2)) {
```
[In button.c](https://github.com/mwpenny/portal64-still-alive/blob/672d4b558d46a3ce1418785e935dd2f0ee63d4a0/src/scene/button.c#L112)

Checking the sleepFrames of the colliding rigid body (e.g. the cube), presents an issue if a trigger is present in the same area as the button, since triggers reset the `sleepFrames` of colliding bodies to `IDLE_SLEEP_FRAMES` in each frame:

```
    if (triggerType & listener->usedTriggerMask) {
        // an object activating a signal should not sleep
        objectEnteringTrigger->body->sleepFrames = IDLE_SLEEP_FRAMES;
    }
```
[In trigger_listener.c](https://github.com/mwpenny/portal64-still-alive/blob/672d4b558d46a3ce1418785e935dd2f0ee63d4a0/src/scene/trigger_listener.c#L46-L49)

I implemented a workaround, adding a pressFrames countdown in `button` that works the same as using `sleepFrames` would.

Reply:
For reference, how far the trigger reaches in the original Portal:

![button_explode](https://github.com/mwpenny/portal64-still-alive/assets/2451901/0fcf218e-be37-4f7a-8e23-30d026b23026)
![near_trigger](https://github.com/mwpenny/portal64-still-alive/assets/2451901/12e15bd8-0910-4777-ae55-fc6d9a5e1050)

Reply:
> I came across a bug in the button code

Good catch. The stabilization you're talking about is a common technique called debouncing. For example, a physical button can exhibit the same behavior when its contacts first touch.

> I'm not sure yet why rigid bodies inside triggers shouldn't sleep. I'd suspect this could be due to consistency in sending the signals. This could be looked into, if rigid body collisions ever get too resource hungry in areas with triggers.

You have the right idea. As an optimization, objects which haven't moved in a while "go to sleep" and stop being checked for collision. But trigger collision need to happen every frame that an object is inside of it, so the signal remains active the whole time. A good way to think about signals is similar to those in electronics.

Given the size of triggers and small number of objects that the player can put inside of them, I'm not too worried about performance here. As the project progresses, profiling will reveal such issues.

> To fix the bug, I implemented a workaround, adding a pressFrames countdown in button that works the same as using sleepFrames would.

This looks good. One question - you added a check for `other->body->flags` before entering the onpress logic. This seems unnecessary as . Did you encounter problems here?

Reply:
> Given the size of triggers and small number of objects that the player can put inside of them, I'm not too worried about performance here. As the project progresses, profiling will reveal such issues.

Agreed. I was thinking mostly about whole room encompassing triggers like in chamber13, just something to keep in mind.

> One question - you added a check for `other->body->flags` before entering the onpress logic. This seems unnecessary . Did you encounter problems here?

You're correct, that check is unnecessary. It's just a remnant from testing different things that somehow slipped through, I corrected that now.

Reply:
Thanks

> I was thinking mostly about whole room encompassing triggers like in chamber13

That's true. What I meant by the second half of what I said is that even in these cases, there are typically only 1 or 2 cubes. Good to keep an eye on like you said.

Reply:
This is nice too since it lets the signal fire while still holding the cube, which the original game does but Portal 64 didn't before.